src/java/org/apache/cassandra/service/accord/AccordTracing.java

```diff
@@ -149,176 +749,265 @@ private static long uniqueNowMicros()
         {
-            Event event;
+            TxnEvent event;
 
             @Override
-            public EnumMap<TraceEventType, TraceState> apply(TxnId id, EnumMap<TraceEventType, TraceState> cur)
+            public TxnEvents apply(TxnId id, TxnEvents state)
             {
-                if (cur == null)
+                if (state == null)
                     return null;
 
-                TraceState curState = cur.get(eventType);
-                if (curState == null || curState.permits == 0)
-                    return cur;
-
-                if (count.incrementAndGet() >= MAX_EVENTS)
-                {
-                    count.decrementAndGet();
-                    noSpamLogger.warn("Too many Accord trace events stored already; delete some to continue tracing");
-                }
-                else
-                {
-                    curState.permits--;
-                    curState.addInternal(event = new Event());
-                }
-                return cur;
+                event = state.trace(kind, globalCount);
+                return state;
             }
         }
         Register register = new Register();
-        stateMap.compute(txnId, register);
+        txnIdMap.compute(txnId, register);
         return register.event;
     }
 
-    public void setPermits(TxnId txnId, TraceEventType eventType, int newPermits)
+    // null values, or values < 0, are ignored
+    public void set(TxnId txnId, CoordinationKinds trace, BucketMode newBucketMode, int newBucketSize, int newBucketSubSize, int newBucketSeen, float newChance, boolean unsetManagedByPattern)
     {
-        stateMap.compute(txnId, (id, cur) -> {
-            if (newPermits != 0)
-            {
-                if (cur == null)
-                    cur = new EnumMap<>(TraceEventType.class);
-                cur.computeIfAbsent(eventType, ignore -> new TraceState()).permits = newPermits;
-            }
-            else if (cur != null)
+        Invariants.requireArgument(newBucketSize != 0);
+        Invariants.requireArgument(newBucketSubSize != 0);
+        Invariants.requireArgument(Float.isNaN(newChance) || (newChance <= 1.0f && newChance > 0f));
+        txnIdMap.compute(txnId, (id, cur) -> {
+            if (cur == null)
             {
-                TraceState curState = cur.get(eventType);
-                if (curState != null)
-                {
-                    if (!curState.isEmpty()) curState.permits = 0;
-                    else
-                    {
-                        cur.remove(eventType);
-                        if (cur.isEmpty())
-                            return null;
-                    }
-                }
+                if (newBucketSize < 0)
+                    throw new IllegalArgumentException("Must specify bucket size for new trace config.");
+
+                cur = new TxnEvents();
+                if (newBucketSubSize < 0)
+                    cur.bucketSubSize = newBucketSize;
```

Comment:
Should we always set bucket sub size to new bucket size if `newBucketjsubSize < 0`, and not only when `cur == null`?

In other words, move this if outside of `cur == null` if statement?